File: solutions/exercise-09/src/main/java/io/miragon/training/application/service/ClaimMembershipService.java
package io.miragon.training.application.service;

import io.miragon.training.application.port.inbound.ClaimMembershipUseCase;
import io.miragon.training.application.port.outbound.MembershipRepository;
import io.miragon.training.domain.MembershipId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.concurrent.atomic.AtomicInteger;

@Service
@Transactional
public class ClaimMembershipService implements ClaimMembershipUseCase {

    private static final Logger log = LoggerFactory.getLogger(ClaimMembershipService.class);

    private static final int MAX_SPOTS = 1000;
    private final AtomicInteger claimedSpots = new AtomicInteger(0);

    private final MembershipRepository repository;

    public ClaimMembershipService(MembershipRepository repository) {
        this.repository = repository;
    }

    @Override
    public boolean claimMembership(MembershipId membershipId) {
        var membership = repository.find(membershipId);
        var claimed = claimedSpots.incrementAndGet();
        var hasSpot = claimed <= MAX_SPOTS;
        log.info("Claiming membership for {} – spot {}/{}, hasSpot={}", membership.email().value(), claimed, MAX_SPOTS, hasSpot);
        return hasSpot;
    }
}

File: solutions/exercise-4/src/main/kotlin/io/miragon/training/application/service/ClaimMembershipService.kt
package io.miragon.training.application.service

import io.github.oshai.kotlinlogging.KotlinLogging
import io.miragon.training.application.port.inbound.ClaimMembershipUseCase
import io.miragon.training.application.port.outbound.MembershipRepository
import io.miragon.training.domain.MembershipId
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.concurrent.atomic.AtomicInteger

@Service
@Transactional
class ClaimMembershipService(
    private val repository: MembershipRepository,
) : ClaimMembershipUseCase {

    private val log = KotlinLogging.logger {}

    private val claimedSpots = AtomicInteger(0)
    private val maxSpots = 10

    override fun claimMembership(membershipId: MembershipId): Boolean {
        val membership = repository.find(membershipId)
        val claimed = claimedSpots.incrementAndGet()
        val hasSpot = claimed <= maxSpots
        log.info { "Claiming membership for ${membership.email.value} – spot $claimed/$maxSpots, hasSpot=$hasSpot" }
        return hasSpot
    }
}
